qiskit/algorithms/optimizers/gradient_descent.py

```diff
@@ -12,17 +12,46 @@
 
 """A standard gradient descent optimizer."""
 
-from typing import Iterator, Optional, Union, Callable, Dict, Any, List, Tuple
-from functools import partial
-
+from dataclasses import dataclass, field
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List, Iterator
 import numpy as np
-
 from .optimizer import Optimizer, OptimizerSupportLevel, OptimizerResult, POINT
+from .steppable_optimizer import AskData, TellData, OptimizerState, SteppableOptimizer
+from .utils.learning_rate import LearningRate
 
 CALLBACK = Callable[[int, np.ndarray, float, float], None]
 
 
-class GradientDescent(Optimizer):
+@dataclass
+class GradientDescentState(OptimizerState):
+    """State of :class:`~.GradientDescent`.
+
+    Dataclass with all the information of an optimizer plus the learning_rate and the stepsize.
+    """
+
+    stepsize: Optional[float]
+    """Norm of the gradient on the last step."""
+    learning_rate: LearningRate = field(compare=False)
```

Comment:
I was looking at the docs for this. I have no idea why the formatting is off. Maybe the = field(... has some effect.

![image](https://user-images.githubusercontent.com/40241007/185242565-bda23291-df70-4a0c-93f8-35eb24ca4b0e.png)

The main reason I was looking is how would a user know, reading the API ref, that with the constructor something other than a LearningRate type can be given? And since the extra values are something that can be passed when building a Learning rate, so if you have x you do LearningRate(x) to pass it in it seems simple enough. I am not exactly opposed to having this I just wonder the net real value especially if its not surfaced in any way in the docs.

Also at present, it maybe due to the typo, the types that learning_rate supports are not coming out when I go there

![image](https://user-images.githubusercontent.com/40241007/185244306-da0acefe-4a30-41dc-813e-7febcda7a31e.png)

Reply:
It is not the =field(...) nor the typo. I have tried changing the order of the attributes and the _init_ file and that is not it either.
I am still looking into it.

Reply:
On an other note, should I just add a note to the attribute learning_rate of gradient descent saying that in case that the user inputs something which is not a LearningRate it will try to convert it to a learning rate, or should I just forget about it and delete the post_init?

Reply:
At present ` inputs something which is not a LearningRate` means giving it on the constructor right - if you were to directly update/access the field it has to/will be of type LearningRate. For me it always seemed that if I had one of these other types, that you auto-convert in post_init there, all that it is saving me is that instead of passing say `x` I have to pass `LearningRate(x)` which never seemed that much of a hardship but kept consistency with the constructor and direct variable access of that field in respect of types.

I think you have a better idea of how its going to be used etc. If you think that's its of real value to have that autoconvert support then we need to document the support provided by your post_init. If you decide to drop that and go only with LearningRate, then as a user I have to build a LearningRate explicitly which seems simple enough. Now if you go this way, and later it was felt desirable to allow the extra types, say we got feedback from users or whatever, and do that LearningRate constructions on their behalf of in a post init like you have it there it could always be added back etc later if it seemed really needed.

Reply:
Okay, in that case I think it would be better to remove the post_init. In any case, a non-heavy user won't have to deal with it since `GradientDescentState` is typically created by `GradientDescent` which already takes some generator and converts it into `LearningRate` internally.

Reply:
Okay, I have looked into it in detail and talked to @manoelmarques and it seems that we are going to have to live with it. The issue seems to come from sphinx not properly documenting the typehints of attributes in dataclasses. This problem is also present in some of nature's documentation.

Reply:
Actually the one difference I will note about that field that is not great with the docs is that it has an underscore in its name - the others that come out do not have. Indeed when its rendered there the underscore is gone too from it!

Reply:
In the screenshot you sent it also says `learning_rate` with an underscore. However it is covered by the name of the class of the type hint in capital letters. That is why what we see is `LEARNINGRATEe:` . `e:` is just the end of `learning_rate:` which has one more character because of the underscore. I don't know if you were refering to that one not having an underscore or to something else?